### PR TITLE
Added function to calculate cumulative GPA(not including current semester)

### PR DESCRIPTION
**Pull request by insertcustomname**
Added a function to calculate cumulative GPA excluding current GPA. Doesn't display cumulative currently.

Signed-off-by: Suhas Hariharan hariharan774531@sas.edu.sg

### CODE REVIEW

**Comment by gary-kim** [comment]
I'll take a better look at this when I have more time but from a quick cursory look, some simplification of the code would be nice. Some functions better documented, async/await use, and use of classes. Otherwise, looks pretty good. Thank you for this.

**Comment by insertcustomname** [comment]
Sure, I just added another quick change to have a basic UI, when the button is pressed it calculates and shows the current cumulative gpa, excluding the current semester if the term has not finished.

**Comment by gary-kim** [comment]
Also, could you amend your commits to include a DCO in all of them?

File: src/js/saspowerschoolff.js [hunk 1]
```diff
@@ -143,14 +144,15 @@ function main_page () {
...
-                            link: $prev_course.getElementsByTagName("td")[2].getElementsByTagName("a")[0].href,
                             fp: -1,
 
                         });
                     }
                     $("table[border='0'][cellpadding='3'][cellspacing='1'][width='100%']").prepend(`<tr><td align="center">Last Semester GPA (S1): ${calculate_gpa(courses_first_semester)}</td></tr>`);
                 }
             });
+        $("table[border='0'][cellpadding='3'][cellspacing='1'][width='100%']").prepend(`<td style="background-color: white;" align="center"><button id="calculateCumulative">Calculate Cumulative GPA</button></td>`);
+        $("#calculateCumulative").click(function () { show_cumulative_gpa(); });
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
        $("#calculateCumulative").click(show_cumulative_gpa);
```
No need to make a function just to call another function :).

File: src/js/saspowerschoolff.js [hunk 2]
```diff
@@ -188,6 +196,73 @@ function login_page () {
...
+        .then(data => {
+            const el = document.createElement("html");
+            el.innerHTML = data;
+            const tabs = el.getElementsByClassName("tabs")[0].getElementsByTagName("li");
+            for (let i = 0; i < tabs.length; i++) {
+                fetches.push(
+                    fetch(tabs[i].getElementsByTagName("a")[0].href)
+                        .then(function (response) {
+                            return response.text();
+                        })
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
                        .then(res => res.text())
```

File: src/js/saspowerschoolff.js [hunk 3]
```diff
@@ -188,6 +196,73 @@ function login_page () {
...
+                            element_list = semester.getElementsByClassName("box-round")[0].getElementsByTagName("table")[0];
+                            element_list = element_list.getElementsByTagName("tbody")[0].getElementsByTagName("tr");
+                            for (let t = 2; t < element_list.length; t++) {
+                                if (element_list[t].innerText.trim() === ("S2")) {
+                                    all_courses.push(courses);
+                                    courses = [];
+                                }
+                                if (element_list[t].getElementsByTagName("th").length > 0) {
+                                    continue;
+                                } else {
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]: No need for the else here.

File: src/js/saspowerschoolff.js [hunk 4]
```diff
@@ -188,6 +196,73 @@ function login_page () {
...
+                                    const $prev_course = element_list[t];
+                                    courses.push({
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]: Using the Course class would be nice.

**Comment by insertcustomname** [comment]
> Also, could you amend your commits to include a DCO in all of them?
> EDIT 1: I think this could be better as a Vue component but if you'd rather keep it as is for now, that's fine. We can push that as technical debt and someone else can make a PR to migrate that to Vue.
> EDIT 2: Could you use the [Course](https://github.com/gary-kim/saspes/blob/master/src/js/models/Course.js) class? In this case, you'd probably have to add a new optional variable to the class for `creditHours`.
> EDIT 3: This is more a note for the Vue migration than now but it'd be nice to have a loading indicator to indicate that the information is being processed. That's complex to do in this case but should be far easier with Vue.

I tried to modify the old commit but was unable to, is there a git command I can use to add a signoff to all old commits?

It probably would work better as a Vue component, but perhaps we can just keep it like this for now and migrate it to Vue later? I'm not really familiar with Vue so unfortunately I can't do that just yet. 

I'll modify it to use the Course class.

**Comment by gary-kim** [comment]
You can run `git rebase HEAD~6 --signoff` then run `git push --force-with-lease origin cumulative-gpa`. That should add a signoff to all your commits here.

**Comment by insertcustomname** [comment]
Do you think it's okay to merge now @gary-kim?

**Comment by insertcustomname** [comment]
I have an idea for calculating overall cumulative that I just want to check with you @gary-kim, what if it checks the array of courses from the current semester against the latest in the grade history? If they differ it knows it's a different semester and adds it to the average, else it assumes it's the same semester and ignores that.

**Comment by insertcustomname** [comment]
Alternatively we can check the current date against the end of the grading period, according to the &enddate parameter that date is june 30th, are you aware if this is true or not? i.e the date when the current semester's grades would show up in powerschool.

**Comment by gary-kim** [comment]
> I have an idea for calculating overall cumulative that I just want to check with you @gary-kim, what if it checks the array of courses from the current semester against the latest in the grade history? If they differ it knows it's a different semester and adds it to the average, else it assumes it's the same semester and ignores that.

Sounds like an okay idea. Another option is, it looks like the grade links do not include dates if that period is over. You can see from looking at the links for semester 1. Could that be an option?

In any case, I think we can move that back into a cleanup PR. This PR could use some better code comments but other than that, looks good.

**Comment by insertcustomname** [comment]
Sorry, not particularly sure how to do that, how would I go about doing that?

**Comment by gary-kim** [comment]
You'd use `git fetch` then `git rebase origin/master`, then follow the given instructions. The issue is with your branch that there are a lot of conflicts. I just tried it and ran into 4 conflicts that had to be manually addressed.

Would you rather I just push up the already rebased branch?

**Comment by insertcustomname** [comment]
I tried to do that, followed all the steps, did it work or did I mess it up further?

**Comment by gary-kim** [comment]
Looks like it got a bit messed up. A function got duplicated and some stuff that was removed on the master branch got readded and there's now several merge commits.
BTW, you're going to have to force push rebases with `git push --force`.

**Comment by insertcustomname** [comment]
I think I fixed that now.

**Comment by gary-kim** [comment]
Now there's more commits. We were trying to reduce the number of commits :).

**Comment by insertcustomname** [comment]
> You'd use `git fetch` then `git rebase origin/master`, then follow the given instructions. The issue is with your branch that there are a lot of conflicts. I just tried it and ran into 4 conflicts that had to be manually addressed.
> 
> Would you rather I just push up the already rebased branch? I can also use the opportunity to squash your commits.

Yep, I haven't really figured out this git stuff sorry, can we just revert to a previous commit and do what you mentioned here?

**Comment by insertcustomname** [comment]
For future reference if I add something on a different branch and the master changes in the meanwhile what should I do? Just pull from master and run git rebase master?

**Comment by gary-kim** [comment]
> For future reference if I add something on a different branch and the master changes in the meanwhile what should I do? Just pull from master and run git rebase master?

That will do. You may want to use `git rebase origin/master` because `git rebase master` will rebase on top of your local master which may not be up to date with the remote master. Some projects use other merge methods which means it's fine to use merges from master but we use merge commits here so a lot of people say it's preferable not to have merges from master.

I'll do a final review when I get some time. This should be the first thing in after `v0.20.0` gets released.

**Comment by insertcustomname** [comment]
If I make any changes(i.e to add comments and clean it up more), should I do that in a separate pull request or just add it here?

**Comment by gary-kim** [comment]
> If I make any changes(i.e to add comments and clean it up more), should I do that in a separate pull request or just add it here?

Probably better to add it here.

**Comment by insertcustomname** [comment]
I tried to squash my commits to add comments and somehow accidentally included a commit from a different branch here, is it okay if I just close this pull request and make a new one with a single commit?

**Comment by gary-kim** [comment]
You can always clean up commits and force push them onto the branch. At worst, you can delete the branch locally, recreate everything, then force push that back onto the remote branch. In any case, looks like you already created another branch.

**Pull request closed without merging**